v3_async_wip/v3_async_wip/iothub_session.py

```diff
@@ -0,0 +1,284 @@
+        *,
+        hostname: str,  # iothub_hostname?
+        device_id: str,
+        module_id: Optional[str] = None,
+        ssl_context: Optional[ssl.SSLContext] = None,
+        shared_access_key: Optional[str] = None,
+        sastoken_fn: Optional[custom_typing.FunctionOrCoroutine] = None,
+        **kwargs,
+    ) -> None:
+        """
+        :param str device_id: The device identity for the IoT Hub device containing the
+            IoT Hub module
+        :param str module_id: The module identity for the IoT Hub module
+        :param str hostname: Hostname of the IoT Hub or IoT Edge the device should connect to
+        :param ssl_context: Custom SSL context to be used by the client
+            If not provided, a default one will be used
+        :type ssl_context: :class:`ssl.SSLContext`
+        :param str shared_access_key: A key that can be used to generate SAS Tokens
+        :param sastoken_fn: A function or coroutine function that takes no arguments and returns
+            a SAS token string when invoked
+
+        :raises: ValueError if none of 'ssl_context', 'symmetric_key' or 'sastoken_fn' are provided
+        :raises: ValueError if both 'symmetric_key' and 'sastoken_fn' are provided
+        :raises: ValueError if an invalid 'symmetric_key' is provided
+        :raises: TypeError if an invalid keyword argument is provided
+        """
+        # Validate parameters
+        _validate_kwargs(**kwargs)
+        if shared_access_key and sastoken_fn:
+            raise ValueError(
+                "Incompatible authentication - cannot provide both 'shared_access_key' and 'sastoken_fn'"
+            )
+        if not shared_access_key and not sastoken_fn and not ssl_context:
+            raise ValueError(
+                "Missing authentication - must provide one of 'shared_access_key', 'sastoken_fn' or 'ssl_context'"
+            )
+
+        # Set up SAS auth (if using)
+        generator: Optional[st.SasTokenGenerator]
+        # NOTE: Need to keep a reference to the SasTokenProvider so we can stop it during cleanup
+        self._sastoken_provider: Optional[st.SasTokenProvider]
+        if shared_access_key:
+            uri = _format_sas_uri(hostname=hostname, device_id=device_id, module_id=module_id)
+            signing_mechanism = sm.SymmetricKeySigningMechanism(shared_access_key)
+            generator = st.InternalSasTokenGenerator(signing_mechanism=signing_mechanism, uri=uri)
+            self._sastoken_provider = st.SasTokenProvider(generator)
+        elif sastoken_fn:
+            generator = st.ExternalSasTokenGenerator(sastoken_fn)
+            self._sastoken_provider = st.SasTokenProvider(generator)
+        else:
+            self._sastoken_provider = None
+
+        # Create a default SSLContext if not provided
+        if not ssl_context:
+            ssl_context = _default_ssl_context()
+
+        # Instantiate the MQTTClient
+        cfg = config.IoTHubClientConfig(
+            hostname=hostname,
```

Comment:
Nit: inconsistent naming -- you're shortening the variable name instead of the module name.

Reply:
Good call, this was just me being lazy